pkg/envoy/agent.go

```diff
@@ -180,9 +182,15 @@ func (a *Agent) terminate() {
 			default:
 				if err != nil {
 					log.Errorf(err.Error())
-					a.abortCh <- errAbort
-					log.Infof("Graceful termination logic ended prematurely, error while obtaining downstream_cx_active stat")
-					break graceful_loop
+					retryCount++
```

Comment:
instead if complicating this loop why do not we add this in `activeProxyConnections` method? Similar to how we do [here](https://github.com/istio/istio/blob/f3c6047d3984b169aebf4ad7bebe944ac1309ba2/security/pkg/nodeagent/cache/secretcache.go#L429)

Reply:
> And statusCh exits from envoy closing, which would certainly make activeProxyConnections fail?

Yes.

Reply:
In terms of separation of responsibilities, I think @ramaraochavali  suggestions are good, but Adding an exponential backoff to the ActiveConnection method doesn't seem to make a difference to the sleep holding the ticker loop, because inside the ticker loop we're waiting to get that output synchronously.

Reply:
Were able to test when you mentioned exponential backoff does not seem to make a difference?

Reply:
Also,. if this is reproducible, can you please enable debug/trace logs and share?

Reply:
ok, I haven't tried turning it on before, so I can't confidently say it can handle all workloads of our service, but I'll try to make it reproducible as much as possible.

Reply:
>Were able to test when you mentioned exponential backoff does not seem to make a difference? 

i mean 

```
for range ticker.C {
  ac, err := a.activeProxyConnections() // <- if it has exponential backoff, it make this 1s ticker loop blocked during retry 
  select {
  case status := <- a.statusCh: 
  ...
  default 
}
```

is it ok if that block select sentence?

Reply:
I see what you mean. As I mentioned at other place, retry without exponential backoff might not be that useful. But let us wait for your log when you are able to reproduce.

Reply:
hard to activate for all workloads ( loki, log storage would under load), so I've activated it for some workloads. I'll share it with you as soon as it's reproduced

Reply:
As a result, if retry with backoff is necessary to solve the problem, how about changing the structure to something like this?

Separation of api thread and result processing thread:

```
        // envoy stats loop
        activeConnChan := make(chan int)
	quitActiveConnLoop := make (chan bool)
	go func() {
		for {
			select {
			case <-quitActiveConnLoop:
				return
			default:
				activeProxyConnWithChannel(activeConnChan) // with exponential backoff retry 
				time.Sleep(2 * time.Second) // ticker period example
			}
		}
	}()
	
	// result processing loop
	for {
		select {
		case ac := <-activeConnChan:
			if ac == 0 {
				//do something
				quitActiveConnLoop <- true
				
			}
			if ac == -1 {
				//do something
				quitActiveConnLoop <- true
			}
		case status := <-a.statusCh:
			// do something
			quitActiveConnLoop <- true
			return
		}
	}
```